kymatio/scattering2d/tests/test_torch_scattering2d.py

```diff
@@ -0,0 +1,376 @@
+                for j2 in range(j1 + 1, J):
+                    for l2 in range(L):
+                        n += 1
+                        order2.append(n)
+        assert len(order0) == n_order0
+        assert len(order1) == n_order1
+        assert len(order2) == n_order2
+        return order0, order1, order2
+
+    @pytest.mark.parametrize("backend", backends)
+    @pytest.mark.parametrize("device", devices)
+    def test_Scattering2D(self, backend, device):
+        test_data_dir = os.path.dirname(__file__)
+        data = torch.load(os.path.join(test_data_dir, 'test_data_2d.pt'))
+
+        x = data['x']
+        S = data['Sx']
+        J = data['J']
+
+        # we need to reorder S from interleaved (how it's saved) to o0, o1, o2
+        # (which is how it's now computed)
+
+        o0, o1, o2 = self.reorder_coefficients_from_interleaved(J, L=8)
+        reorder = torch.from_numpy(np.concatenate((o0, o1, o2)))
+        S = S[..., reorder, :, :]
+
+        pre_pad = data['pre_pad']
+
+        M = x.shape[2]
+        N = x.shape[3]
+
+        # Then, let's check when using pure pytorch code
+        scattering = Scattering2D(J, shape=(M, N), pre_pad=pre_pad, backend=backend, frontend='torch')
+        Sg = []
+        x = x.to(device)
+        scattering.to(device)
+        S = S.to(device)
+        if backend.name == 'torch_skcuda' and device == 'cpu':
+            with pytest.raises(TypeError) as ve:
+                Sg = scattering(x)
+            assert "cuda" in ve.value.args[0]
+        else:
+            Sg = scattering(x)
+            assert torch.allclose(Sg, S)
+
+        # check also the default backend
+        scattering = Scattering2D(J, shape=(M, N), pre_pad=pre_pad, frontend='torch')
+        Sg = []
+        x = x.to(device)
+        scattering.to(device)
+        S = S.to(device)
+        Sg = scattering(x)
+        assert torch.allclose(Sg, S)
+
+    @pytest.mark.parametrize("backend", backends)
+    def test_batch_shape_agnostic(self, backend):
+        J = 3
+        L = 8
+        shape = (32, 32)
+
+        shape_ds = tuple(n // 2 **J for n in shape)
+
+        S = Scattering2D(J, shape, L, backend=backend, frontend='torch')
+
+        with pytest.raises(RuntimeError) as ve:
+            S(torch.zeros(()))
+        assert "at least two" in ve.value.args[0]
+
+        with pytest.raises(RuntimeError) as ve:
+            S(torch.zeros((32, )))
+        assert "at least two" in ve.value.args[0]
+
+        x = torch.zeros(shape)
+
+        if backend.name == 'torch_skcuda':
+            x = x.cuda()
+            S.cuda()
+
+        Sx = S(x)
+
+        assert len(Sx.shape) == 3
+        assert Sx.shape[-2:] == shape_ds
+
+        n_coeffs = Sx.shape[-3]
+
+        test_shapes = ((1,) + shape, (2,) + shape, (2, 2) + shape,
+                       (2, 2, 2) + shape)
+
+        for test_shape in test_shapes:
+            x = torch.zeros(test_shape)
+
+            if backend.name == 'torch_skcuda':
```

Comment:
https://github.com/kymatio/kymatio/pull/415/files#r321476299 May we continue this discussion here?

@edouardoyallon @janden .

Reply:
Sorry I'm confused the link isn't working: the one about shape dimensions?

Reply:
@janden said: Why aren't we doing S.to(device) here?

@edouardoyallon responded with: in order to test lines which are ignored otherwise because some functionality are not implemented.

@janden do you think that the convo is resolved?

Reply:
Yeah - it just increases the coverage...

Reply:
indeed

Reply:
Ok will change, then.

Reply:
Never mind. I'm not sure what I was thinking when I wrote that. Will leave as is.